src/custom-elements/src/components/carousel/carousel.tsx

```diff
@@ -0,0 +1,227 @@
+import { Component, Element, Event, EventEmitter, Prop, State, h } from '@stencil/core';
+import { ACTIVE_CLASS, CAROUSEL_CLASSES, TRANSITIONING_CLASS, UTILITY_CLASSES } from '../../constants/classes';
+import { ANIMATION_DURATION } from '../../constants/constants';
+import { ThreeStepsAnimation } from '../../utils/ThreeStepsAnimation';
+
+@Component({
+  tag: 'chi-carousel',
+  styleUrl: 'carousel.scss',
+  scoped: true
+})
+export class Carousel {
+  @Element() el: HTMLElement;
+  /**
+   * To render Carousel with dot controllers
+   */
+  @Prop() dots: boolean;
+  /**
+   * To render Carousel with a single item per view
+   */
+  @Prop() single: boolean;
+  /**
+   *  View
+   */
+  @State() view = 0;
+  /**
+   *  Number of views
+   */
+  @State() numberOfViews = 0;
+  @State() customPrevButton: boolean;
+  @State() customNextButton: boolean;
+  /**
+   * Triggered when the user navigates to another view
+   */
+  @Event() chiViewChange: EventEmitter<number>;
+
+  private wrapper?: HTMLElement;
+  private scrollBreakpoints = {
+    0: 50
+  };
+  private fullScrollLength: number;
+  private animation: ThreeStepsAnimation;
+  private _animationClasses: string;
+  private resizeHandler = () => {
+    this.fullScrollLength = this.calculateWidth(this.el);
+
+    const wrapperWidth = this.calculateWidth(this.wrapper);
+
+    this.numberOfViews = Math.ceil(wrapperWidth / this.fullScrollLength);
+
+    const remainder = wrapperWidth - (this.fullScrollLength * this.numberOfViews);
+
+    if (this.single) {
+      this._applySizeToItems();
+    }
+    this.calculateScrollBreakpoints(this.fullScrollLength, wrapperWidth, remainder);
+  };
+
+  calculateWidth(element: HTMLElement) {
+    return element.offsetWidth;
+  }
+
+  calculateScrollBreakpoints(fullScrollLength: number, wrapperWidth: number, remainder: number) {
```

Comment:
perhaps it'd be a good idea to add jsdoc to this function so that other devs can understand better all these calculations?